Sources/Algorithms/FirstRange.swift

```diff
@@ -0,0 +1,133 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// firstRange(of:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  @inlinable
+  public func firstRange<Other: Collection>(
+    of other: Other,
+    by areEquivalent: (Element, Other.Element) throws -> Bool
+  ) rethrows -> Range<Index>? {
+    var searchStart = startIndex
+    
+    guard let needleFirst = other.first else {
+      return searchStart..<searchStart
+    }
+
+    while let matchStart = try self[searchStart...]
+            .firstIndex(where: { try areEquivalent($0, needleFirst) })
+    {
+      var index = matchStart
+      var otherIndex = other.startIndex
+
+      repeat {
+        formIndex(after: &index)
+        other.formIndex(after: &otherIndex)
+        
+        if otherIndex == other.endIndex {
+          return matchStart..<index
+        } else if index == endIndex {
+          return nil
+        }
+      } while try areEquivalent(self[index], other[otherIndex])
+
+      searchStart = self.index(after: matchStart)
+    }
+
+    return nil
+  }
+}
+
+extension Collection where Element: Equatable {
+  @inlinable
+  public func firstRange<Other: Collection>(of other: Other) -> Range<Index>?
+    where Other.Element == Element
+  {
+    firstRange(of: other, by: ==)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// lastRange(of:)
+//===----------------------------------------------------------------------===//
+
+extension BidirectionalCollection {
+  @inlinable
+  public func lastRange<Other: BidirectionalCollection>(
+    of other: Other,
+    by areEquivalent: (Element, Other.Element) throws -> Bool
+  ) rethrows -> Range<Index>? {
+    var searchEnd = endIndex
```

Comment:
I originally thought this method would just use wrapping `ReversedCollection`s, call `firstRange(of:)`, and counter-convert the returned indexes.  I don't know how much of an improvement this customized code is.  (The downside is that improvements to `firstRange(of:)` aren't automatically inherited.)  Maybe we could use benchmarking?

Reply:
That would have been nice, but `ReversedCollection` unfortunately does an index decrement inside its [subscript](https://github.com/apple/swift/blob/main/stdlib/public/core/Reverse.swift#L246-L249), which makes it unsuitable for such purposes.